File: apps/android/shared/src/main/java/com/eatfair/shared/data/remote/ChatService.kt
package com.eatfair.shared.data.remote

import com.eatfair.shared.config.AppConfig
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Chat Service
 * Matches iOS ChatService.swift for platform parity
 * Real-time messaging via WebSocket
 */

// Data Models for Real-time Chat (prefixed to avoid conflict with ApiModels.ChatMessage)
data class RealtimeChatConversation(
    val id: String,
    val orderId: String? = null,
    val rideId: String? = null,
    val customerId: String,
    val driverId: String,
    val customerName: String,
    val driverName: String,
    val lastMessage: String? = null,
    val lastMessageAt: Long? = null,
    val unreadCount: Int = 0,
    val createdAt: Long
)

data class RealtimeChatMessage(
    val id: String,
    val conversationId: String,
    val senderId: String,
    val senderType: RealtimeSenderType,
    val messageType: RealtimeMessageType,
    val content: String,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val isRead: Boolean = false,
    val createdAt: Long
)

enum class RealtimeSenderType {
    CUSTOMER,
    DRIVER
}

enum class RealtimeMessageType {
    TEXT,
    LOCATION,
    IMAGE,
    QUICK_REPLY
}

data class RealtimeChatEvent(
    val type: RealtimeChatEventType,
    val conversationId: String,
    val message: RealtimeChatMessage? = null,
    val data: Map<String, Any> = emptyMap()
)

enum class RealtimeChatEventType {
    NEW_MESSAGE,
    MESSAGE_READ,
    TYPING,
    STOPPED_TYPING,
    CONNECTION_STATUS
}

sealed class RealtimeChatResult<out T> {
    data class Success<T>(val data: T) : RealtimeChatResult<T>()
    data class Error(val message: String) : RealtimeChatResult<Nothing>()
}

@Singleton
class ChatService @Inject constructor() {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private var webSocket: WebSocket? = null

    // Use AppConfig for environment-aware URLs (production)
    private val baseUrl: String
        get() = AppConfig.API_BASE_URL.removeSuffix("/api")

    private val wsUrl: String
        get() {
            val httpUrl = AppConfig.API_BASE_URL.removeSuffix("/api")
            return httpUrl.replace("http://", "ws://").replace("https://", "wss://")
        }

    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _chatEvents = MutableSharedFlow<RealtimeChatEvent>()
    val chatEvents: Flow<RealtimeChatEvent> = _chatEvents

    private val _messages = MutableStateFlow<List<RealtimeChatMessage>>(emptyList())
    val messages: StateFlow<List<RealtimeChatMessage>> = _messages.asStateFlow()

    enum class ConnectionState {
        DISCONNECTED,
        CONNECTING,
        CONNECTED,
        ERROR
    }

    // Quick reply templates
    val customerQuickReplies = listOf(
        "I'm at pickup",
        "Running late",
        "Can you call me?",
        "Thank you!",
        "Where are you?",
        "I'm coming down"
    )

    val driverQuickReplies = listOf(
        "On my way!",
        "I've arrived",
        "Share your location?",
        "I'll wait here",
        "Be there in 5 min",
        "Having trouble finding you"
    )

    /**
     * Create a new conversation
     */
    suspend fun createConversation(
        orderId: String? = null,
        rideId: String? = null,
        customerId: String,
        driverId: String,
        token: String
    ): RealtimeChatResult<RealtimeChatConversation> {
        return try {
            val json = JSONObject().apply {
                orderId?.let { put("order_id", it) }
                rideId?.let { put("ride_id", it) }
                put("customer_id", customerId)
                put("driver_id", driverId)
            }

            val requestBody = json.toString().toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("$baseUrl/api/chat/conversations")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post(requestBody)
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val conversation = parseConversation(JSONObject(responseBody ?: "{}"))
                RealtimeChatResult.Success(conversation)
            } else {
                RealtimeChatResult.Error("Failed to create conversation: ${response.message}")
            }
        } catch (e: Exception) {
            RealtimeChatResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Send a text message
     */
    suspend fun sendMessage(
        conversationId: String,
        content: String,
        senderType: RealtimeSenderType,
        token: String
    ): RealtimeChatResult<RealtimeChatMessage> {
        return try {
            val json = JSONObject().apply {
                put("content", content)
                put("sender_type", senderType.name.lowercase())
                put("message_type", "text")
            }

            val requestBody = json.toString().toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("$baseUrl/api/chat/conversations/$conversationId/messages")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post(requestBody)
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val message = parseMessage(JSONObject(responseBody ?: "{}"))
                // Add to local messages
                _messages.value = _messages.value + message
                RealtimeChatResult.Success(message)
            } else {
                RealtimeChatResult.Error("Failed to send message: ${response.message}")
            }
        } catch (e: Exception) {
            RealtimeChatResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Send location message
     */
    suspend fun sendLocation(
        conversationId: String,
        latitude: Double,
        longitude: Double,
        senderType: RealtimeSenderType,
        token: String
    ): RealtimeChatResult<RealtimeChatMessage> {
        return try {
            val json = JSONObject().apply {
                put("content", "Shared location")
                put("sender_type", senderType.name.lowercase())
                put("message_type", "location")
                put("latitude", latitude)
                put("longitude", longitude)
            }

            val requestBody = json.toString().toRequestBody("application/json".toMediaType())

            val request = Request.Builder()
                .url("$baseUrl/api/chat/conversations/$conversationId/messages")
                .addHeader("Authorization", "Bearer $token")
                .addHeader("Content-Type", "application/json")
                .post(requestBody)
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val message = parseMessage(JSONObject(responseBody ?: "{}"))
                _messages.value = _messages.value + message
                RealtimeChatResult.Success(message)
            } else {
                RealtimeChatResult.Error("Failed to send location: ${response.message}")
            }
        } catch (e: Exception) {
            RealtimeChatResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Fetch message history
     */
    suspend fun fetchMessages(
        conversationId: String,
        token: String,
        limit: Int = 50,
        before: Long? = null
    ): RealtimeChatResult<List<RealtimeChatMessage>> {
        return try {
            var url = "$baseUrl/api/chat/conversations/$conversationId/messages?limit=$limit"
            before?.let { url += "&before=$it" }

            val request = Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer $token")
                .get()
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                val responseBody = response.body?.string()
                val jsonArray = JSONArray(responseBody ?: "[]")
                val messages = mutableListOf<RealtimeChatMessage>()
                for (i in 0 until jsonArray.length()) {
                    messages.add(parseMessage(jsonArray.getJSONObject(i)))
                }
                _messages.value = messages
                RealtimeChatResult.Success(messages)
            } else {
                RealtimeChatResult.Error("Failed to fetch messages: ${response.message}")
            }
        } catch (e: Exception) {
            RealtimeChatResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Mark messages as read
     */
    suspend fun markAsRead(
        conversationId: String,
        token: String
    ): RealtimeChatResult<Unit> {
        return try {
            val request = Request.Builder()
                .url("$baseUrl/api/chat/conversations/$conversationId/read")
                .addHeader("Authorization", "Bearer $token")
                .post("".toRequestBody(null))
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                // Update local messages to mark as read
                _messages.value = _messages.value.map { it.copy(isRead = true) }
                RealtimeChatResult.Success(Unit)
            } else {
                RealtimeChatResult.Error("Failed to mark as read: ${response.message}")
            }
        } catch (e: Exception) {
            RealtimeChatResult.Error(e.message ?: "Unknown error")
        }
    }

    /**
     * Connect to WebSocket for real-time messaging
     */
    fun connectWebSocket(conversationId: String, token: String) {
        _connectionState.value = ConnectionState.CONNECTING

        val request = Request.Builder()
            .url("$wsUrl/ws/chat/$conversationId")
            .addHeader("Authorization", "Bearer $token")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _connectionState.value = ConnectionState.CONNECTED
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val json = JSONObject(text)
                    val event = parseChatEvent(json)
                    _chatEvents.tryEmit(event)

                    // Add new message to local list
                    event.message?.let { message ->
                        _messages.value = _messages.value + message
                    }
                } catch (e: Exception) {
                    // Handle parse error
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _connectionState.value = ConnectionState.ERROR
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
                _connectionState.value = ConnectionState.DISCONNECTED
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _connectionState.value = ConnectionState.DISCONNECTED
            }
        })
    }

    /**
     * Send typing indicator via WebSocket
     */
    fun sendTypingIndicator(isTyping: Boolean) {
        val json = JSONObject().apply {
            put("type", if (isTyping) "typing" else "stopped_typing")
        }
        webSocket?.send(json.toString())
    }

    /**
     * Disconnect WebSocket
     */
    fun disconnectWebSocket() {
        webSocket?.close(1000, "User disconnected")
        webSocket = null
        _connectionState.value = ConnectionState.DISCONNECTED
    }

    /**
     * Clear local messages
     */
    fun clearMessages() {
        _messages.value = emptyList()
    }

    private fun parseConversation(json: JSONObject): RealtimeChatConversation {
        return RealtimeChatConversation(
            id = json.optString("id", ""),
            orderId = json.optString("order_id", null),
            rideId = json.optString("ride_id", null),
            customerId = json.optString("customer_id", ""),
            driverId = json.optString("driver_id", ""),
            customerName = json.optString("customer_name", "Customer"),
            driverName = json.optString("driver_name", "Driver"),
            lastMessage = json.optString("last_message", null),
            lastMessageAt = if (json.has("last_message_at")) json.optLong("last_message_at") else null,
            unreadCount = json.optInt("unread_count", 0),
            createdAt = json.optLong("created_at", System.currentTimeMillis())
        )
    }

    private fun parseMessage(json: JSONObject): RealtimeChatMessage {
        return RealtimeChatMessage(
            id = json.optString("id", ""),
            conversationId = json.optString("conversation_id", ""),
            senderId = json.optString("sender_id", ""),
            senderType = RealtimeSenderType.valueOf(
                json.optString("sender_type", "CUSTOMER").uppercase()
            ),
            messageType = RealtimeMessageType.valueOf(
                json.optString("message_type", "TEXT").uppercase()
            ),
            content = json.optString("content", ""),
            latitude = if (json.has("latitude")) json.optDouble("latitude") else null,
            longitude = if (json.has("longitude")) json.optDouble("longitude") else null,
            isRead = json.optBoolean("is_read", false),
            createdAt = json.optLong("created_at", System.currentTimeMillis())
        )
    }

    private fun parseChatEvent(json: JSONObject): RealtimeChatEvent {
        val typeString = json.optString("type", "NEW_MESSAGE")
        val type = try {
            RealtimeChatEventType.valueOf(typeString.uppercase().replace(" ", "_"))
        } catch (e: Exception) {
            RealtimeChatEventType.NEW_MESSAGE
        }

        val message = if (json.has("message")) {
            parseMessage(json.getJSONObject("message"))
        } else null

        return RealtimeChatEvent(
            type = type,
            conversationId = json.optString("conversation_id", ""),
            message = message
        )
    }
}

File: apps/ios/eatfair-ios-shared/Sources/EatFairShared/Services/ChatService.swift
import Foundation
import Combine
import os

private let logger = Logger(subsystem: "com.dollorai.shared", category: "ChatService")

/// Dollor.ai Chat Service
// TODO: [CRITICAL] API mismatch — Double URL prefix (chatServiceURL includes /api/chat, then paths add /api/chat again). Also path structure differs from backend (/api/chat/send vs /api/chat/conversations/{id}/messages). All 6 endpoints will 404.
/// Real-time messaging between independent drivers and customers
/// MATCHMAKING PLATFORM - Facilitates communication, does not monitor content
public class ChatService: ObservableObject {
    public static let shared = ChatService()

    // MARK: - Configuration
    private var baseURL: String {
        return AppConfig.shared.chatServiceURL
    }

    @Published public var isLoading = false
    @Published public var error: String?
    @Published public var messages: [ChatMessage] = []
    @Published public var activeConversation: Conversation?

    private var cancellables = Set<AnyCancellable>()
    private var webSocket: URLSessionWebSocketTask?

    private init() {}

    // MARK: - Chat Models

    public enum MessageType: String, Codable {
        case text = "text"
        case location = "location"
        case quickReply = "quick_reply"
        case system = "system"
    }

    public struct ChatMessage: Codable, Identifiable {
        public let id: String
        public let conversationId: String
        public let senderId: String
        public let senderType: String
        public let messageType: MessageType
        public let content: String
        public let locationLat: Double?
        public let locationLng: Double?
        public let isRead: Bool
        public let createdAt: String

        enum CodingKeys: String, CodingKey {
            case id = "message_id"
            case conversationId = "conversation_id"
            case senderId = "sender_id"
            case senderType = "sender_type"
            case messageType = "message_type"
            case content
            case locationLat = "location_lat"
            case locationLng = "location_lng"
            case isRead = "is_read"
            case createdAt = "created_at"
        }
    }

    public struct Conversation: Codable, Identifiable {
        public let id: String
        public let rideId: String?
        public let orderId: String?
        public let customerId: String
        public let driverId: String?
        public let isActive: Bool
        public let lastMessageAt: String?
        public let createdAt: String

        enum CodingKeys: String, CodingKey {
            case id = "conversation_id"
            case rideId = "ride_id"
            case orderId = "order_id"
            case customerId = "customer_id"
            case driverId = "driver_id"
            case isActive = "is_active"
            case lastMessageAt = "last_message_at"
            case createdAt = "created_at"
        }
    }

    public struct QuickReply: Identifiable {
        public let id = UUID()
        public let text: String
        public let category: String
    }

    // MARK: - Quick Replies

    public static let customerQuickReplies: [QuickReply] = [
        QuickReply(text: "I'm at the pickup location", category: "location"),
        QuickReply(text: "Running a few minutes late", category: "timing"),
        QuickReply(text: "Can you call me?", category: "contact"),
        QuickReply(text: "Thank you!", category: "gratitude")
    ]

    public static let driverQuickReplies: [QuickReply] = [
        QuickReply(text: "On my way!", category: "status"),
        QuickReply(text: "I've arrived", category: "location"),
        QuickReply(text: "Can you share your location?", category: "location"),
        QuickReply(text: "I'll wait here", category: "status")
    ]

    // MARK: - Create Conversation

    /// Start a new conversation for a ride or delivery
    public func createConversation(
        customerId: String,
        driverId: String? = nil,
        rideId: String? = nil,
        orderId: String? = nil,
        completion: @escaping (Result<Conversation, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/chat/conversations") else {
            completion(.failure(ChatError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: Any] = ["customer_id": customerId]
        if let driverId = driverId { body["driver_id"] = driverId }
        if let rideId = rideId { body["ride_id"] = rideId }
        if let orderId = orderId { body["order_id"] = orderId }

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        isLoading = true
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.isLoading = false

                if let error = error {
                    self?.error = error.localizedDescription
                    completion(.failure(error))
                    return
                }

                guard let data = data else {
                    completion(.failure(ChatError.noData))
                    return
                }

                do {
                    let conversation = try JSONDecoder().decode(Conversation.self, from: data)
                    self?.activeConversation = conversation
                    completion(.success(conversation))
                } catch {
                    self?.error = "Failed to decode conversation: \(error.localizedDescription)"
                    completion(.failure(error))
                }
            }
        }.resume()
    }

    // MARK: - Send Message

    /// Send a text message
    public func sendMessage(
        conversationId: String,
        senderId: String,
        senderType: String,
        content: String,
        completion: @escaping (Result<ChatMessage, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/chat/conversations/\(conversationId)/messages") else {
            completion(.failure(ChatError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "sender_id": senderId,
            "sender_type": senderType,
            "message_type": "text",
            "content": content
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        performMessageRequest(request, completion: completion)
    }

    // MARK: - Send Location

    /// Share current location
    public func sendLocation(
        conversationId: String,
        senderId: String,
        senderType: String,
        latitude: Double,
        longitude: Double,
        completion: @escaping (Result<ChatMessage, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/chat/conversations/\(conversationId)/messages") else {
            completion(.failure(ChatError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "sender_id": senderId,
            "sender_type": senderType,
            "message_type": "location",
            "content": "Shared location",
            "location_lat": latitude,
            "location_lng": longitude
        ]

        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        performMessageRequest(request, completion: completion)
    }

    // MARK: - Fetch Messages

    /// Get message history for a conversation
    public func fetchMessages(
        conversationId: String,
        limit: Int = 50,
        offset: Int = 0,
        completion: @escaping (Result<[ChatMessage], Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/chat/conversations/\(conversationId)/messages?limit=\(limit)&offset=\(offset)") else {
            completion(.failure(ChatError.invalidURL))
            return
        }

        isLoading = true
        URLSession.shared.dataTask(with: url) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.isLoading = false

                if let error = error {
                    self?.error = error.localizedDescription
                    completion(.failure(error))
                    return
                }

                guard let data = data else {
                    completion(.failure(ChatError.noData))
                    return
                }

                do {
                    struct MessagesResponse: Codable {
                        let messages: [ChatMessage]
                    }
                    let response = try JSONDecoder().decode(MessagesResponse.self, from: data)
                    self?.messages = response.messages
                    completion(.success(response.messages))
                } catch {
                    self?.error = "Failed to decode messages: \(error.localizedDescription)"
                    completion(.failure(error))
                }
            }
        }.resume()
    }

    // MARK: - Mark as Read

    /// Mark messages as read
    public func markAsRead(
        conversationId: String,
        readerId: String,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/chat/conversations/\(conversationId)/read") else {
            completion(.failure(ChatError.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = ["reader_id": readerId]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    completion(.failure(error))
                    return
                }
                completion(.success(()))
            }
        }.resume()
    }

    // MARK: - WebSocket Connection

    /// Connect to real-time chat updates
    public func connectWebSocket(conversationId: String, participantType: String, participantId: String) {
        let wsURLString = baseURL.replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://", with: "ws://")
        guard let url = URL(string: "\(wsURLString)/ws/chat/\(conversationId)?participant_type=\(participantType)&participant_id=\(participantId)") else {
            return
        }

        let session = URLSession(configuration: .default)
        webSocket = session.webSocketTask(with: url)
        webSocket?.resume()

        receiveMessage()
    }

    private func receiveMessage() {
        webSocket?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    if let data = text.data(using: .utf8),
                       let chatMessage = try? JSONDecoder().decode(ChatMessage.self, from: data) {
                        DispatchQueue.main.async {
                            self?.messages.append(chatMessage)
                        }
                    }
                case .data(let data):
                    if let chatMessage = try? JSONDecoder().decode(ChatMessage.self, from: data) {
                        DispatchQueue.main.async {
                            self?.messages.append(chatMessage)
                        }
                    }
                @unknown default:
                    break
                }
                self?.receiveMessage()
            case .failure(let error):
                #if DEBUG
                logger.info("[ChatService] WebSocket error: \(error)")
                #endif
            }
        }
    }

    /// Send message via WebSocket
    public func sendWebSocketMessage(content: String) {
        let messageDict: [String: Any] = [
            "type": "message",
            "content": content
        ]
        if let data = try? JSONSerialization.data(withJSONObject: messageDict),
           let jsonString = String(data: data, encoding: .utf8) {
            webSocket?.send(.string(jsonString)) { error in
                if let error = error {
                    #if DEBUG
                    logger.info("[ChatService] WebSocket send error: \(error)")
                    #endif
                }
            }
        }
    }

    /// Disconnect WebSocket
    public func disconnectWebSocket() {
        webSocket?.cancel(with: .goingAway, reason: nil)
        webSocket = nil
    }

    // MARK: - Helper Methods

    private func performMessageRequest(_ request: URLRequest, completion: @escaping (Result<ChatMessage, Error>) -> Void) {
        isLoading = true
        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            DispatchQueue.main.async {
                self?.isLoading = false

                if let error = error {
                    self?.error = error.localizedDescription
                    completion(.failure(error))
                    return
                }

                guard let data = data else {
                    completion(.failure(ChatError.noData))
                    return
                }

                do {
                    let message = try JSONDecoder().decode(ChatMessage.self, from: data)
                    self?.messages.append(message)
                    completion(.success(message))
                } catch {
                    self?.error = "Failed to decode message: \(error.localizedDescription)"
                    completion(.failure(error))
                }
            }
        }.resume()
    }
}

// MARK: - Error Types

public enum ChatError: LocalizedError {
    case invalidURL
    case noData
    case conversationNotFound
    case notAuthorized

    public var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .noData:
            return "No data received"
        case .conversationNotFound:
            return "Conversation not found"
        case .notAuthorized:
            return "Not authorized to access this conversation"
        }
    }
}
